### PR TITLE
fix(readiness): detect DGX Station hardware independently

### LINKED ISSUE #10906: DGX Station Express rejects the qualified Colossus BaseOS display-name variant
## Investigation summary

- A DGX Station GB300 reports `DGX_PRETTY_NAME="NVIDIA DGX GB300WS"`, `DGX_SWBUILD_VERSION="7.5.0-GB300ws-GB200ws"`, and `DGX_SWBUILD_DATE="2026-04-02-08-20-16"`.
- `--force-station-install` completes Station host preparation, but topology discovery and onboarding classify the same release marker again and stop with `DGX Station hardware or software profile is not qualified.`
- Current `main` at `0673b122147433e8025222a135c7b8a3ebdbd27a` reproduces the release-classification failure.
- Both current classifiers recognize the same version and build date as `supported-colossus-baseos` only when `DGX_PRETTY_NAME="NVIDIA DGX Server"`.
- No open issue matches this exact metadata variant.

## Description

Station Express rejects an observed display-name variant of the already qualified April 2026 Colossus BaseOS profile. The installer’s temporary metadata override reaches `STATION_FACTORY_HOST_READY`, but later readiness and onboarding checks reparse `/etc/dgx-release` without an override contract and reject the profile.

The expected result is automatic recognition of this exact qualified profile variant. Unknown release metadata must remain fail-closed.

## Reproduction steps

1. Use a DGX Station GB300 on Ubuntu 24.04 with:

   ```text
   DGX_PRETTY_NAME="NVIDIA DGX GB300WS"
   DGX_SWBUILD_VERSION="7.5.0-GB300ws-GB200ws"
   DGX_SWBUILD_DATE="2026-04-02-08-20-16"
   DGX_PLATFORM="DGX Server for GALAXY-GB300"
   ```

2. Run:

   ```bash
   curl -fsSL https://www.nvidia.com/nemoclaw.sh | bash -s -- --force-station-install
   ```

3. Continue with either the default Nemotron Ultra selection or `--station-deepseek`.
4. Observe that host preparation succeeds, but topology discovery or onboarding stops before model or sandbox creation.
5. Run the suggested `nemoclaw onboard --resume --name my-assistant` command and observe the same failure.

## Environment

- Hardware: DGX Station GB300
- OS: Ubuntu 24.04
- Reported install result: NemoClaw v0.0.109
- Current-source reproduction: NemoClaw `main` commit `0673b122147433e8025222a135c7b8a3ebdbd27a`
- GPU, driver, ECC, Docker, CDI, disk, and factory-runtime checks: reported healthy

## Debug output

```text
This DGX Station OS image is outside the recognized Station Express release-metadata boundary.
STATION_FACTORY_HOST_READY
DGX Station hardware or software profile is not qualified.
```

No model or sandbox is created.

## Proposed design

Decision: Accept

Reason and placement: Core NemoClaw. Extend the existing Colossus BaseOS release profile in `scripts/prepare-dgx-station-host.sh` and `src/lib/readiness/platform-qualification.ts`. Accept both reviewed display names for the exact `7.5.0-GB300ws-GB200ws` and `2026-04-02-08-20-16` tuple. Do not propagate `--force-station-install` into general readiness or direct-GPU policy checks.

Accountable maintainer: @ericksoa

Validation plan:

- Add focused classifier tests for both accepted display names.
- Keep build-version, build-date, platform, unsafe-marker, and near-match variants denied.
- Add readiness regression coverage for the reported tuple.
- Run the focused Station installer and readiness tests.
- Update the Station preparation documentation.
- Provide DGX Station evidence for the tested commit before approval, as repository policy requires.

This design keeps the release marker fail-closed and uses the existing `supported-colossus-baseos` owner. It does not add override state, an environment-variable bypass, or another policy authority.

## Related work

- #7124 accepted the exact April 2026 Colossus BaseOS profile and requires separate approval for metadata variants.
- #7132 added the temporary metadata-only installer override.
- #9898 established that reviewed DGX display-name variants can share one qualified release family while other fields remain constrained.
- #7979 resolved a different 7.5.0 factory-image tuple through exact profile recognition.

## Acceptance criteria

- Both current Station classifiers return `supported-colossus-baseos` for either reviewed display name when version, build date, and platform match.
- Version, build-date, platform, marker-safety, and near-match drift remain denied.
- Host readiness qualifies the reported profile only when the existing Station GB300 hardware, Ubuntu 24.04, ARM64, PCI, GPU, and runtime requirements pass.
- Stock DGX OS, NVIDIA AI Developer Tools, and generic Ubuntu behavior does not change.
- The canonical Station preparation page describes both accepted display names for this exact profile.
- Reviewable DGX Station evidence identifies the tested commit, profile, result, and supporting link before approval.

## Checklist

- [x] I confirmed this bug is reproducible
- [x] I searched existing issues and this is not a duplicate

### LINKED ISSUE #10928: Converge DGX Station qualification on stable identity signals
## Outcome

Make DGX Station qualification tolerate cosmetic release-label changes without accepting a new hardware generation or software build family.

NemoClaw must preserve every currently qualified Station profile. `DGX_PRETTY_NAME` must remain diagnostic evidence instead of deciding Station hardware or software lineage.

## Current problem

Station qualification uses two release classifiers and multiple firmware-product matchers. Both release classifiers have repeatedly rejected qualified hosts when NVIDIA changed or omitted a descriptive field.

Recent examples include:

- #7224: an OTA-upgraded host omitted `DGX_OTA_PRETTY_NAME`.
- #7979: an AI Developer Tools image used another qualified build date.
- #9898: stock DGX OS used another `DGX_PRETTY_NAME` value.
- #10906: the qualified Colossus image used the same display-name variant in a second profile.

The current host checks already obtain stronger Station evidence from firmware, the exact GB300 PCI identity, ARM64 Ubuntu, `DGX_PLATFORM`, release-version structure, and runtime probes.

Operation and failure class: read-only DGX Station platform qualification; false negative caused by cosmetic metadata and duplicated classifier behavior.

## Product decision

Decision: Accept

Reason and placement: Core NemoClaw. Converge the Station release contract across the shell installer and TypeScript readiness path. Stop using `DGX_PRETTY_NAME` as an authoritative identity field. Preserve all existing hardware, software-family, runtime, and action-safety gates.

Accountable maintainer: @ericksoa

Validation plan: Run one shared corpus through both production classifiers. Preserve all current positive profiles and negative near matches. Run the focused Station installer, readiness, onboarding, and documentation tests. Obtain exact-commit DGX Station evidence before approval because the host-preparation classifier is security-sensitive.

This decision is consistent with the accepted readiness direction in #7407 and the Station Express delivery scope in #8379.

### Hardware identity

Keep the existing fail-closed requirements:

- Linux ARM64 on Ubuntu 24.04.
- Qualified DGX Station GB300 firmware identity.
- A display-class NVIDIA PCI device with GB300 device ID `0x31c2` or `0x31c3`.
- An effective NVIDIA GPU.

### Release classification

- Keep `/etc/dgx-release` file ownership, file type, size, schema, duplicate-field, and OTA-pair validation unchanged.
- Keep `DGX_PLATFORM="DGX Server for GALAXY-GB300"` mandatory for every DGX release profile.
- Keep the accepted OTA versions, no-OTA `7.6.x` family, and exact factory build-version and build-date boundaries unchanged.
- Keep `DGX_OTA_PRETTY_NAME="DGX OS"` authoritative when that optional lineage field is present.
- Do not require a specific `DGX_PRETTY_NAME` value after hardware, platform, version, OTA shape, and factory-build identity qualify.
- Continue parsing `DGX_PRETTY_NAME` as bounded inert metadata, but do not use it for qualification.

### Current positive corpus

Both production classifiers must continue to accept:

- Generic Ubuntu 24.04 ARM64 on qualified Station GB300 hardware.
- OTA DGX OS `7.2.0`, `7.4.0`, and `7.5.0`, with the accepted full-image and OTA-upgrade marker forms.
- No-OTA DGX OS numeric `7.6.x` with both observed display names.
- April 2026 Colossus BaseOS with both observed display names.
- May and June 2026 AI Developer Tools images.
- Both accepted GB300 PCI device IDs.

### Negative corpus

Both classifiers must continue to reject:

- Unsafe, malformed, oversized, or ambiguous release markers.
- A different `DGX_PLATFORM`.
- Partial or malformed OTA history.
- Unsupported OTA versions and future no-OTA release families.
- Factory-image version or build-date drift.
- Non-Station firmware, non-GB300 hardware, unsupported OS, and unsupported architecture.

## Implementation constraints

- Use one shared behavior corpus to assert shell and TypeScript classifier parity.
- Do not add a runtime network lookup or remote profile catalog.
- Do not propagate `--force-station-install` into general readiness.
- Do not broaden package, service, driver, Docker, CDI, or direct-GPU mutation authority.
- Do not classify arbitrary future Colossus or AI Developer Tools builds as supported.
- Keep unknown or contradictory evidence fail-closed with actionable diagnostics.

## Acceptance criteria

- [ ] Every current positive corpus entry returns the same qualified profile from both classifiers.
- [ ] Every current negative corpus entry returns the same denied result from both classifiers.
- [ ] Changing only `DGX_PRETTY_NAME` does not change the result for an otherwise qualified profile.
- [ ] Changing platform, release version, OTA structure, factory build date, marker safety, firmware identity, or PCI identity still denies qualification.
- [ ] Installer detection, host preparation, topology selection, onboarding, resume, and direct-GPU policy consume the same qualification result.
- [ ] Public Station documentation describes display names as diagnostic values rather than qualification requirements.
- [ ] Exact-commit physical Station evidence confirms one accepted profile and one denied near match before approval.

## Non-goals

- Automatic support for unqualified future Station image families.
- A durable metadata-bypass receipt.
- Changes to Station package installation or factory-runtime repair.
- Changes to the supported-platform or managed-inference matrices.

## Related work

- #7407
- #7410
- #7411
- #8379
- #8382
- #10906
- #10914

### PR DESCRIPTION
**Pull request by ericksoa**
## Outcome

NemoClaw identifies DGX Station GB300 hardware independently from descriptive firmware and release labels.
Changing `DGX_PRETTY_NAME` no longer rejects a genuine Station, while unknown software, conflicting identity evidence, unsupported runtime state, and missing exact GB300 PCI evidence still fail closed.

Fixes #10906
Fixes #10928

## Architecture

- `src/lib/inference/dgx-station-identity.ts` owns bounded firmware classification, exact GB300 PCI identity, conflict detection, and the shared Python probe fragment.
- The pre-Node Station preparation helper implements the same hardware contract for installer entry.
- Readiness reports separate hardware, software, runtime, and combined Station qualifications.
- Installer selection, local NIM detection, direct-GPU policy, and both dual-Station paths consume that contract instead of inventing their own display-name rules.
- Product family, board, device tree, system vendor, CPU count, host memory, GPU count and memory, base OS, and DGX release values are sanitized diagnostics. They do not authorize Station behavior.

## Safety boundaries

- Hardware admission requires a bounded Station GB300 firmware value plus NVIDIA vendor `0x10de`, GB300 device `0x31c2` or `0x31c3`, and an exact display-class PCI value.
- Conflicting NVIDIA firmware identities and incomplete or oversized evidence fail closed.
- Station Express still requires Linux ARM64, Ubuntu 24.04, an effective NVIDIA GPU, and a qualified software profile.
- `--force-station-install` remains explicit validation-only handling for a trusted, complete, but unrecognized release profile. It stops before onboarding.
- Direct-GPU policy creation and dual-Station planning use the same qualified identity.

## Review scope

The hardware/software split remains within the accepted scope of #10928. The pre-Node architecture redesign remains deferred to #11133. Aaron Erickson's repository role was verified as `admin` through GitHub's collaborator-permission API.

## Verification

Current candidate: `d83597c642eac4c63d10c9b22881e777a7547a79`.

- Both installer entry points stop when either Station classifier fails. The original failing-helper regression remains covered.
- Bash and TypeScript now use the same ASCII firmware-token grammar. The non-ASCII separator and case-folding regressions failed before the fix and pass afterward.
- Validation-only mode rejects `--station-deepseek`, `NEMOCLAW_VLLM_MODEL`, and `NEMOCLAW_MODEL` before continuing through either installer entry point.
- The public bootstrap help documents `--force-station-install` and its no-onboarding boundary.
- Fifteen shared hardware cases run the actual shell helper and TypeScript classifiers against product, family, board, device-tree, malformed/control-character, firmware-size, missing-PCI, and PCI-scan-bound fixtures.
- All 31 classifier, entrypoint, discovery, and public-help checks passed. All 232 focused identity, readiness, policy, and dual-Station source tests passed. The full Express installer suite passed with its one existing skip.
- CLI build, repository checks, ShellCheck, secret scanning, and `npm run validate:pr` passed against canonical base `1b3cd3668c0530969de8a24e941587793a0ea7c4`. No validator configuration or audit policy was relaxed.
- GitHub verifies all 33 PR commits. CI and automatic review results are attached to the current candidate.

## Physical Station evidence

[The operator's physical DGX Station test record](https://github.com/NVIDIA/NemoClaw/pull/10914#issuecomment-5595398080) reports two passes at `0b33c0e4511dbe50dc29997950381835e75d12ee`:

- Read-only hardware classification and `--check` on a Dell Pro Max with Station GB300, Ubuntu 24.04.4 ARM64, and the actual April 2026 Colossus BaseOS marker: `READY_FOR_FACTORY_RUNTIME_PREPARATION`.
- A controlled copy of that release marker with `DGX_PRETTY_NAME="NVIDIA DGX GB300WS"`: `supported-colossus-baseos` and `GB300WS_VARIANT_RESULT=PASS`.

The tested `scripts/prepare-dgx-station-host.sh` is byte-for-byte unchanged in this candidate (Git blob `c40997a611ed3e19127e10404db34b7da9a6cbcd`). The actual release file was not modified during the variant test.

The physical record covers preflight and release classification. It does not claim a physical denied-near-match run, TypeScript readiness execution, `--apply`, full onboarding, model serving, or sandbox creation. The checked-in parity and denial tests provide deterministic negative-boundary evidence.

---
Signed-off-by: Aaron Erickson <aerickson@nvidia.com>

## Summary by CodeRabbit

- **New Features**
  - DGX Station qualification now verifies firmware, device-tree, and PCI hardware evidence.
  - Readiness results separately report Station hardware, software, runtime, and combined qualification.
  - Express setup provides clearer handling for unsupported, conflicting, or incomplete Station identities.
  - WSL Express selection now chooses the appropriate local or managed inference runtime.

- **Bug Fixes**
  - Improved rejection of malformed, conflicting, or unsupported platform metadata.
  - Expanded support for recognized DGX Station release and identity variants.

- **Documentation**
  - Clarified validation-only recovery, supported release requirements, and Station installation boundaries.

### CODE REVIEW

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

File: docs/reference/platform-support.mdx [hunk 1]
```diff
@@ -79,7 +79,7 @@ For install requirements and the shorter setup-oriented platform view, refer to
...
...ice`. Capacity and display values remain diagnostic evidence and do not replace firmware and PCI hardware identity. Clean-host end-to-end validation passed on generic Ubuntu and Colossus BaseOS; stock DGX OS and AI Developer Tools completed Station Express validation. The DGX OS `7.5.0` run used released OpenShell `0.0.85`, local Nemotron Ultra serving, sandbox `cuInit(0)`, and a Hermes write/read file-tool task. A dual-Station configuration has not been validated, and dedicated CI coverage is not available. Direct-GPU policies expose only the exact read-only BDF directory for each discovered display-class PCI device with NVIDIA vendor ID (`0x10de`) and GB300 device ID (`0x31c2` or `0x31c3`) plus required existing topology and module paths; they do not expose `/sys`, the PCI parent subtree, or sysfs write access. During physical validation, reads of `/sys/fs/cgroup/cgroup.controllers` and `/sys/class/net/lo/address` remained denied. For canonical hardware qualification, image requirements, preparation, repair limits, reboot handoff, and the explicit temporary metadata override, see [Prepare DGX Station to Install NemoClaw](../get-started/additional-setup/dgx-station-preparation). |
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply docs/reference/platform-support.mdx]: The cited physical-validation statements predate this PR (they are present on main from bbdd85008c3) and describe earlier profile validation. This PR does not claim physical validation for its changed qualification logic: the PR remains draft and its body requires exact-commit Station evidence before readiness or approval. The added row text only documents the new hardware/software split and validation-only boundary, so the historical claims are retained.

File: test/installer-integration/station-release-classifier-parity.test.ts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+      noOta("7.5.0-GB300ws-GB200ws", "2026-04-02-08-20-16").replace(
+        DISPLAY_NAME,
+        'DGX_PRETTY_NAME="NVIDIA DGX GB300WS"',
+      ),
+    ],
+    [
+      "changed-label Colossus BaseOS",
+      "supported-colossus-baseos",
+      noOta("7.5.0-GB300ws-GB200ws", "2026-04-02-08-20-16"),
+    ],
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply test/installer-integration/station-release-classifier-parity.test.ts]: Addressed in bed4e4e561: the shared parity table now includes the April 2026 Colossus tuple with DGX_PRETTY_NAME="NVIDIA DGX Server", alongside the GB300WS and workstation labels.

File: test/installer-integration/station-release-classifier-parity.test.ts [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+      `source "$STATION_PREPARE" >/dev/null
+stat() { printf '%s\n' "$MARKER_METADATA"; }
+uname() { printf 'aarch64\n'; }
+station_os_release_path() { printf '%s' "$FIXTURE/os-release"; }
+station_product_name_path() { printf '%s' "$FIXTURE/product_name"; }
+station_product_family_path() { printf '%s' "$FIXTURE/product_family"; }
+station_board_name_path() { printf '%s' "$FIXTURE/board_name"; }
+station_device_tree_model_path() { printf '%s' "$FIXTURE/model"; }
+dgx_station_release_path() { printf '%s' "$FIXTURE/dgx-release"; }
+station_has_exact_gb300_pci_gpu() { return 0; }
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply test/installer-integration/station-release-classifier-parity.test.ts]: Addressed on the current head: exact GB300 PCI evidence is required before every host profile, and bed4e4e561 adds a supported DGX OS profile case that forces the PCI check to fail and asserts the "Expected an NVIDIA GB300 PCI GPU" stop.

File: src/lib/inference/dgx-station-identity.ts [hunk 4]
```diff
@@ -1,11 +1,115 @@
...
+  readFile: (filePath: string) => string,
+  filePath: string,
+  stripNul = false,
+): string | undefined {
+  try {
+    const contents = readFile(filePath);
+    const normalized = (stripNul ? contents.replace(/\0/g, "") : contents).replace(/\n+$/u, "");
+    if (
+      Buffer.byteLength(normalized) > NVIDIA_FIRMWARE_VALUE_MAX_BYTES ||
+      /[\r\n\0]/u.test(normalized)
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply src/lib/inference/dgx-station-identity.ts]: Addressed in bed4e4e561: TypeScript now rejects every C0 control and DEL, the shell classifier uses the same printable-value boundary, the shared Python producer rejects the same range, and the parity suite covers an embedded tab that previously could match Station GB300.

File: src/lib/inference/nim.test.ts [hunk 5]
```diff
@@ -402,11 +402,12 @@ describe("nim", () => {
       }
     }
 
-    it.each(["NVIDIA DGX Station GB300", "DGX-Station", "P3830"])(
+    it.each(["NVIDIA DGX Station GB300", "NVIDIA Station GB300"])(
       "classifies explicit DGX Station identifier %s as station",
       (model) => {
         withFirmwareModel(model, () => {
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply src/lib/inference/nim.test.ts:408]: Addressed in bed4e4e561: withFirmwareModel now returns empty values for product_family, board_name, and the device-tree model, so these tests no longer read runner firmware accidentally.

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

File: scripts/prepare-dgx-station-host.sh [hunk 6]
```diff
@@ -396,13 +439,111 @@ parse_args() {
...
+}
+
+station_firmware_identity() {
+  local LC_ALL=C class="" output=${1:?firmware identity output is required} path recognized="" station_product="" value
+  for path in "$(station_product_name_path)" "$(station_product_family_path)" "$(station_board_name_path)" "$(station_device_tree_model_path)"; do
+    [[ -r "$path" ]] || continue
+    if [[ "$path" == "$(station_device_tree_model_path)" ]]; then
+      value="$(station_device_tree_model_value "$path")" || continue
+    else
+      value="$(head -c 257 "$path" 2>/dev/null || true)"
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply scripts/prepare-dgx-station-host.sh]: Fixed on the current head. `f2346cbc7d` adds `station_bounded_line_value`, which validates the bounded raw DMI bytes and rejects controls before command substitution; `station_firmware_identity` uses it for product name, product family, and board name. `8f34e39a43` makes the regression oracle exercise an embedded NUL specifically in `product_family`. The 37-case production shell/readiness parity suite and all commit/push/synthetic-merge gates pass.

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
Maintainer scope decision for #10928 and this PR: retain the bounded CPU-count, host-memory, GPU-count/memory, system-vendor, OS-name, and DGX-release fields as non-authorizing operator/readiness evidence. They are intentionally consumed in `host.platform.identity` and the private preparation log to make physical DGX Station diagnosis robust across OEM display and release-label variation.

These diagnostics must never replace the conflict-free firmware plus exact GB300 PCI hardware anchor, the qualified software profile, or the runtime gates. The reduction suggestion to remove them as unused is therefore not applicable to the requested scope. Exact-commit physical Station evidence remains the draft/merge gate.

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

**Comment by ericksoa** [comment]
@coderabbitai review

File: scripts/prepare-dgx-station-host.sh [hunk 7]
```diff
@@ -396,13 +440,128 @@ parse_args() {
...
+    last_byte="$(tail -c 1 "$path" 2>/dev/null | od -An -tu1 | tr -d '[:space:]')"
+    [[ "$last_byte" == "10" ]] || return 1
+  fi
+  head -c $((STATION_IDENTITY_VALUE_MAX_BYTES + 1)) "$path" 2>/dev/null | tr -d '\n'
+}
+
+station_device_tree_model_value() {
+  local byte_count control_count last_byte nul_count path=$1
+  byte_count="$(head -c $((STATION_IDENTITY_VALUE_MAX_BYTES + 2)) "$path" 2>/dev/null | wc -c | tr -d '[:space:]')"
+  [[ "$byte_count" =~ ^[0-9]+$ ]] && ((byte_count <= STATION_IDENTITY_VALUE_MAX_BYTES + 1)) || return 1
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply scripts/prepare-dgx-station-host.sh:478]: Fixed on current head `0722aab1f5`. `station_device_tree_model_value` now permits the 257th byte only when the raw value contains the one allowed NUL terminator; otherwise it returns before emission. The existing production shell/readiness parity case uses a 257-byte printable, unterminated device-tree value and passes 39/39 with both consumers denying Station identity. Exact-base `npm run validate:pr` also passes.

**Comment by ericksoa** [comment]
@coderabbitai review

File: scripts/install.sh [hunk 8]
```diff
@@ -4555,63 +4555,73 @@ is_n1x_host() {
...
   fi
-  if [ -r /sys/class/dmi/id/product_name ]; then
-    model="$(cat /sys/class/dmi/id/product_name 2>/dev/null || true)"
-  fi
-  if [ -z "$model" ] && [ -r /sys/firmware/devicetree/base/model ]; then
-    model="$(tr -d '\0' </sys/firmware/devicetree/base/model 2>/dev/null || true)"
-  fi
-  case "$model" in
-    *DGX*Spark*)
+  firmware_state="$(classify_dgx_station_hardware)"
```
Comment (rsliter) [inline scripts/install.sh]: Blocking: this still reproduces CodeRabbit’s earlier fail-closed finding on the current head. `detect_express_platform` is itself called through command substitution. In that context Bash does not preserve `errexit`, so a nonzero classifier status is swallowed here. I replaced `classify_dgx_station_hardware` with a function that returns 42; `preflight_explicit_express_flags` then called `validate_express_platform_boundary` with an empty value and printed `SURVIVED`. Capture this command’s status explicitly and fail before the case or fallback. Apply the same treatment to the release classifier if it can fail in the same nested context, and add an integration regression test proving a failed real helper boundary cannot reach prompting or onboarding.

**Review by rsliter — CHANGES_REQUESTED** [review]
Reviewed commit `7e22b35922066621e946b187b8675f075cdac40f` against base `afb2342666e0fa6b2dbaf23f1e194283b4787363`. The accepted issues establish the hardware and software qualification split, and all required checks currently pass. Focused local validation passed after building the CLI: platform qualification (78 tests), installer Express prompting (94 passed, 1 skipped), firmware and release parity, initial policy, and Station identity tests.

I am requesting changes because CodeRabbit’s earlier fail-closed finding remains reproducible on the current head even though its older thread is marked resolved. When the hardware classifier returns nonzero, the explicit Express preflight reaches platform validation with an empty value and continues. The inline comment includes the reproduction and remediation. This weakens the Station hardware boundary and conflicts with the accepted fail-closed behavior.

Non-blocking: CodeRabbit’s current suggestion to add the May AI Developer Tools tuple to the shared parity table is valid. Both individual classifiers already cover the tuple, but adding it here would keep the cross-consumer corpus complete. If it is not added in this PR, a narrowly scoped follow-up test PR would provide that value.

Approval will also require the repository-mandated exact-commit DGX Station validation evidence for the change to `scripts/prepare-dgx-station-host.sh`. I did not find that evidence attached to this commit.

**Comment by sandl99** [comment]
## DGX Station manual physical test evidence

Two positive checks passed on a physical DGX Station GB300 at commit `0b33c0e4511dbe50dc29997950381835e75d12ee`. This record contains the operator's terminal output.

- Hardware: Dell Pro Max with Station GB300.
- OS: Ubuntu 24.04.4 LTS, ARM64, kernel `6.17.0-1014-nvidia-64k`.
- Actual release marker: `DGX_PRETTY_NAME="NVIDIA DGX Server"`, April 2026 Colossus BaseOS.
- Evidence scope: physical read-only preflight and a controlled release-marker variant test on the same Station.

### 1. Physical host preflight: PASS

The operator ran the candidate helper with `--classify-station-hardware`, followed by `--check` with `NEMOCLAW_STATION_PREP_SUDO_NONINTERACTIVE=1`.

The helper confirmed Station GB300 hardware, accepted `supported-colossus-baseos`, and completed with `CHECK_RESULT=READY_FOR_FACTORY_RUNTIME_PREPARATION`. Factory packages matched the required inventory. The inference GB300 used driver `595.58.03` with corrected and uncorrected volatile ECC counts of zero. No agent, inference, or running Docker workload was present.

Relevant terminal output from 2026-09-09, 03:19 UTC:

```text
tested_commit=0b33c0e4511dbe50dc29997950381835e75d12ee
hardware_classifier=station-gb300
[station-prepare] 2026-09-09T03:19:08Z version=2026-07-20.1 mode=--check log=disabled_read_only
[station-prepare] 2026-09-09T03:19:09Z station_hardware=confirmed firmware_product=Dell Pro Max with Station GB300
[station-prepare] 2026-09-09T03:19:09Z platform=Dell Pro Max with Station GB300 profile=colossus-baseos release=supported-colossus-baseos os=Ubuntu 24.04.4 LTS arch=aarch64 kernel=6.17.0-1014-nvidia-64k
[station-prepare] 2026-09-09T03:19:09Z package_manager=idle phase=initial_Station_package_preflight
[station-prepare] 2026-09-09T03:19:09Z dpkg_audit=clean
[station-prepare] 2026-09-09T03:19:09Z factory_packages=preserved package_and_driver_mutation=disabled
[station-prepare] 2026-09-09T03:19:09Z baseos_packages=exact
[station-prepare] 2026-09-09T03:19:11Z docker_container_baseline_total=0 running=0
[station-prepare] 2026-09-09T03:19:11Z agent_inference_workloads=none vllm_port=8000 free
[station-prepare] 2026-09-09T03:19:11Z running_docker_containers=none action=initial Station host preparation
[station-prepare] 2026-09-09T03:19:11Z gpu_bdf=0009:06:00.0 gpu=NVIDIA GB300 role=inference driver=595.58.03 ecc_corrected=0 ecc_uncorrected=0
[station-prepare] 2026-09-09T03:19:11Z factory_runtime_commands=present
[station-prepare] 2026-09-09T03:19:11Z CHECK_RESULT=READY_FOR_FACTORY_RUNTIME_PREPARATION
```

The helper reported condition-qualified Colossus failed-unit warnings for `cloud-init.service`, `fluent-bit.service`, `fwupd-refresh.service`, and the four SSSD sockets. Those warnings did not prevent completion.

### 2. Controlled GB300WS release-marker variant: PASS

On the same Station and candidate commit, the operator copied `/etc/dgx-release` into a temporary file. Only `DGX_PRETTY_NAME` changed, to `NVIDIA DGX GB300WS`. The test installed the copy with UID/GID `0:0` and mode `0644`, then passed its path to the production `dgx_station_release_state` function from `scripts/prepare-dgx-station-host.sh`.

The Bash test used `set -euo pipefail` and asserted `supported-colossus-baseos` before printing `GB300WS_VARIANT_RESULT=PASS`.

```text
tested_commit=0b33c0e4511dbe50dc29997950381835e75d12ee
DGX_PRETTY_NAME="NVIDIA DGX GB300WS"
DGX_SWBUILD_DATE="2026-04-02-08-20-16"
DGX_SWBUILD_VERSION="7.5.0-GB300ws-GB200ws"
DGX_PLATFORM="DGX Server for GALAXY-GB300"
release_classifier=supported-colossus-baseos
GB300WS_VARIANT_RESULT=PASS
```

This confirms that the candidate's shell release classifier accepts the display-name variant reported in #10906. The real `/etc/dgx-release` remained unchanged. The GB300WS value was controlled test input; it was not the Station's original release label.

### Coverage limits

- The actual `NVIDIA DGX Server` profile also completed `--check` at main commit `de7f565dd062b6f5affe12218ad825c97efee042`. The GB300WS copy was tested only on the PR commit.
- The denied near-match check required by #10928 remains outstanding.
- These runs did not exercise the TypeScript readiness path, validation-only recovery, full Station Express onboarding, model serving, or sandbox creation. No `--apply` or installer run was performed.

This comment supplies the positive physical and classifier evidence for maintainer review. It does not claim that all acceptance or merge requirements are complete.
